rust/sleeper_core/src/datafusion/cast_udf.rs

```diff
@@ -0,0 +1,873 @@
+//! Contains the code to cast primitive types without causing an error.
+//! Values that undergo a narrowing cast are truncated.
+/*
+ * Copyright 2022-2025 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+use arrow::{
+    array::Array,
+    array::AsArray,
+    compute::CastOptions,
+    datatypes::{DataType, Field, FieldRef, Int32Type, Int64Type},
+    util::display::FormatOptions,
+};
+use datafusion::{
+    common::{exec_err, internal_err, plan_datafusion_err, plan_err},
+    error::Result,
+    logical_expr::{
+        ColumnarValue, ReturnFieldArgs, ScalarFunctionArgs, ScalarUDFImpl, Signature, Volatility,
+        interval_arithmetic::Interval, sort_properties::ExprProperties,
+    },
+    scalar::ScalarValue,
+};
+use std::{hash::Hash, sync::Arc};
+
+/// A UDF for performing primitive casting, but without producing errors if a narrowing conversion
+/// would be outside the target type's range. Casts are performed in the "standard" way, i.e. truncating
+/// upper bits on a narrowing conversion.
```

Comment:
Please can we explain why this is necessary in the comment? We could link to the original issue.